src/clj/rems/api/applications_v2.clj

```diff
@@ -0,0 +1,869 @@
+                       :field/placeholder (localization-for :inputprompt item)
+                       :field/optional (:optional item)
+                       :field/options (:options item)
+                       :field/max-length (:maxlength item)})
+                    (:items form))
+        fields (merge-lists-by :field/id
+                               fields
+                               (:form/fields application))
+        description (->> fields
+                         (filter #(= :description (:field/type %)))
+                         first
+                         :field/value)]
+    (assoc application
+           :form/title (:title form)
+           :form/fields fields
+           :application/description description)))
+
+(defn- assoc-resources [application catalogue-items]
+  (let [resources (->> catalogue-items
+                       (map (fn [item]
+                              {:catalogue-item/id (:id item)
+                               :resource/id (:resource-id item)
+                               :resource/ext-id (:resid item)
+                               :catalogue-item/title (assoc (localization-for :title item)
+                                                            :default (:title item))
+                               :catalogue-item/start (:start item)
+                               :catalogue-item/state (keyword (:state item))}))
+                       (sort-by :catalogue-item/id))]
+    (assoc application :application/resources resources)))
+
+(defn- assoc-licenses [application licenses]
+  (let [licenses (->> licenses
+                      (map (fn [license]
+                             (let [type (keyword (:licensetype license))
+                                   content-key (case type
+                                                 :link :license/link
+                                                 :text :license/text)]
+                               {:license/id (:id license)
+                                :license/type type
+                                :license/start (:start license)
+                                :license/end (:end license)
+                                :license/title (assoc (localization-for :title license)
+                                                      :default (:title license))
+                                content-key (assoc (localization-for :textcontent license)
+                                                   :default (:textcontent license))})))
+                      (sort-by :license/id))]
+    (assoc application :application/licenses (merge-lists-by :license/id
+                                                             licenses
+                                                             (:application/licenses application)))))
+
+(defn- assoc-externals [application {:keys [get-form get-catalogue-item get-license get-user]}]
+  (-> application
+      (assoc-form (get-form (:form/id application)))
```

Comment:
I think it would be better to use plain assoc here and have the helper function calculate the value. It's unintuitive to have `assoc-resources` take `catalogue-items` as parameter. Same applies to the other "assocs".